java/src/test/java/ai/rapids/cudf/MemoryBufferTest.java

```diff
@@ -0,0 +1,162 @@
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MemoryBufferTest extends CudfTestBase {
+  @AfterEach
+  void teardown() {
+    if (PinnedMemoryPool.isInitialized()) {
+      PinnedMemoryPool.shutdown();
+    }
+  }
+
+  @Test
+  public void testAddressOutOfBoundsExceptionWhenCopying() {
+    try (HostMemoryBuffer from = HostMemoryBuffer.allocate(16);
+         HostMemoryBuffer to = HostMemoryBuffer.allocate(16)) {
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(-1, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(16, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(0, from, -1, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(0, from, 16, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(0, from, 0, -1, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(0, from, 0, 17, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(1, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBuffer(0, from, 1, 16, Cuda.DEFAULT_STREAM));
+    }
+  }
+
+  @Test
+  public void testAddressOutOfBoundsExceptionWhenCopyingAsync() {
+    try (HostMemoryBuffer from = HostMemoryBuffer.allocate(16);
+         HostMemoryBuffer to = HostMemoryBuffer.allocate(16)) {
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(-1, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(16, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(0, from, -1, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(0, from, 16, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(0, from, 0, -1, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(0, from, 0, 17, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(1, from, 0, 16, Cuda.DEFAULT_STREAM));
+      assertThrows(AssertionError.class, () -> to.copyFromMemoryBufferAsync(0, from, 1, 16, Cuda.DEFAULT_STREAM));
+    }
+  }
+
+  @Test
+  public void testCopyingFromDeviceToDevice() {
+    try (HostMemoryBuffer in = HostMemoryBuffer.allocate(16);
+         DeviceMemoryBuffer from = DeviceMemoryBuffer.allocate(16);
+         DeviceMemoryBuffer to = DeviceMemoryBuffer.allocate(16);
+         HostMemoryBuffer out = HostMemoryBuffer.allocate(16)) {
+      in.setLong(0, 123456789);
+      from.copyFromHostBuffer(in);
+      to.copyFromMemoryBuffer(0, from, 0, 16, Cuda.DEFAULT_STREAM);
```

Comment:
Would be nice if at least one of the tests tried a copy that wasn't at the start of both src and dest.